mesa_23: init at 23.0.1

Atemu opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Atemu committed:
mesa: refactor for multi-version setup

Atemu committed:
mesa_23: init at 23.0.1

Atemu commented:
cc @hw-lunemann

vcunat commented:
Well, _if_ we're going to do this, we should describe some way how to use it, right?  (I think it will have more value later, on stable NixOS)

I know there are options `hardware.opengl.package` (and `.package32`) but they're marked as internal and I'm not sure if switching this way really works without issues.

Atemu committed:
nixos/opengl: add mesaPackage option

andrevmatos commented on pkgs/top-level/all-packages.nix:
```diff
@@ -22225,12 +22225,15 @@ with pkgs;
-  mesa = callPackage ../development/libraries/mesa {
-    llvmPackages = llvmPackages_15;
-    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
+  mesa_22 = darwin.apple_sdk_11_0.callPackage ../development/libraries/mesa/22.nix {
```
why `apple_sdk_11_0` including for linux envs?

Atemu commented:
@vcunat because `package` is also used for switching out the OGL implementation to something else entirely; i.e. AMD's proprietary driver.

Mesa has its own specialities like the `drivers` output.

vcunat commented:
I don't see the difference (yet).  Both is switching to a different libGL (and friends).  If anything, it should fail by default if both internal code sets it (because some other option) and user sets it.

K900 commented:
I think what they're saying is that Mesa needs to specifically choose the `drivers` output and other drivers don't, but I think this shouldn't really be a problem - we can just tell the users to use `mesa.drivers`.

K900 commented:
Not sure I like that. Sounds really easy to break. I really wish we had a generic "provides" mechanism in nixpkgs sometimes...

vcunat commented:
It could be improved fairly easily, but there just a couple particular packages that make sense to pass, and even generally I wouldn't expect `.drivers` to exist commonly.

Atemu commented:
> even generally I wouldn't expect `.drivers` to exist commonly.

That's sort of why I wanted a mesa-specific option.

I get your points but I still think it's better to provide an option that caters towards mesa specifically as an interface to the end-user.

Atemu committed:
mesa: provide patch versions

This allows us to bump the mesa attribute (used by many packages, see
https://github.com/NixOS/nixpkgs/issues/218232) independently from the packages
used by NixOS users.

L911P commented:
you'll want to apply the fix from https://github.com/NixOS/nixpkgs/pull/219802 because otherwise with mesa 23 intel devices won't have 32bit vulkan drivers.

cole-h commented:
> I know there are options `hardware.opengl.package` (and `.package32`) but they're marked as internal and I'm not sure if switching this way really works without issues.

Just as a data point, I've been able to use this "internal" option just fine: https://github.com/cole-h/nixos-config/blob/3d8ba5971794df6422353c5236e7c06bb36428de/hosts/scadrial/modules/rtx4090-compat.nix#L29-L38. It just doesn't appear in `man configuration.nix`.

K900 commented:
23.0.1 is finally out! https://gitlab.freedesktop.org/mesa/mesa/-/tags/mesa-23.0.1

Scrumplex commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,21 +69,48 @@ in
 
+      mesaPackage = mkOption {
+        type = types.package;
+        default = pkgs.mesa;
+        example = "pkgs.mesa_23";
```
```suggestion
        default = pkgs.mesa;
        defaultText = literalExpression "pkgs.mesa";
        example = literalExpression "pkgs.mesa_23";
```

Scrumplex commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,21 +69,48 @@ in
+      };
+      mesaPackage32 = mkOption {
+        type = types.package;
+        default = pkgs.mesa;
+        example = "pkgs.pkgsi686Linux.mesa_23";
```
```suggestion
        default = pkgs.pkgsi686Linux.mesa;
        defaultText = literalExpression "pkgs.pkgsi686Linux.mesa";
        example = literalExpression "pkgs.pkgsi686Linux.mesa_23";
```

Atemu committed:
nixos/opengl: mesa_22 -> mesa_23

K900 commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,21 +69,50 @@ in
         description = lib.mdDoc ''
           The package that provides the OpenGL implementation.
+
+          The default is mesa's drivers which should cover all OpenGL-capable
```
Upstream spells Mesa with a capital M. We should too, probably.

K900 commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,21 +69,50 @@ in
+        defaultText = literalExpression "pkgs.mesa_23";
+        example = literalExpression "pkgs.mesa_22";
+        description = lib.mdDoc ''
+          The mesa driver package used for rendering support on the system.
```
Nit: rendering -> rendering acceleration?

Atemu replied:
If we're splitting hairs, Mesa has non-accelerated rendering support too ;)

K900 replied:
Honestly, I'd probably go for something like "to provide graphics drivers", even.

K900 commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,21 +69,50 @@ in
+        description = lib.mdDoc ''
+          The mesa driver package used for rendering support on the system.
+
+          You should only need to adjust this if you require a newer mesa
```
Nit: not really the case? You can totally want to update for newer features and such.

Atemu replied:
Reasoning is that if you require newer features, you require a new mesa in my view.

K900 reviewed (changes requested):
Some documentation nits.

K900 commented:
I have no idea why Github is not letting me comment on that line, but `hardware.opengl.package` is also set further down in `opengl.nix` and that should probably be removed if we're adding a real default.

Atemu commented:
Good catch!

K900 commented on nixos/modules/hardware/opengl.nix:
```diff
@@ -69,19 +69,48 @@ in
+        defaultText = literalExpression "pkgs.pkgsi686Linux.mesa_23";
+        example = literalExpression "pkgs.pkgsi686Linux.mesa_22";
+        description = lib.mdDoc ''
+          Same as {option}`mesaPackage` but for the 32-bit mesa on 64-bit
```
One last lower case "mesa" here ^^"

K900 reviewed (approved):
LGTM!

vcunat commented:
At some point we might add a bullet into 23.05 release notes, mentioning that you can switch the driver versions.

The editing tasks around mesa releases got way more complicated by this, needing file creation and (possibly) editing a few of them, but I suppose it's OK.  I know I'm a bit late with looking more properly at this; overall I do like this PR.  Thanks.

Atemu commented:
Looking at [22.11's release notes](https://nixos.org/manual/nixos/stable/release-notes.html#sec-release-22.11), something like this could indeed be mentioned. Perhaps even as a candidate for a release highlight.

K900 commented:
I was thinking we should do that in the staging change that actually updates the default.

vcunat commented:
Could be, but I don't see any need to couple these two.

maxeaubrey commented:
Not sure if it's just me/my procrastination on switching to flakes, but I cannot rebuild my system with bf1785d0a91 in.

```
error: attempt to call something which is not a function but a set

       at /home/systems/nixpkgs/nixos/modules/hardware/opengl.nix:102:23:

          101|         '';
          102|         apply = mesa: mesa.drivers or throw "`mesa` package must have a `drivers` output.";
             |                       ^
          103|       };
```

K900 commented:
Can you run that with `--show-trace`? Is `hardware.opengl.package` set in your config?

maxeaubrey commented:
@K900 I do not set `hardware.opengl.package` explicitly anywhere in my config. Sorry about the lack of a full trace, here you go: https://gist.github.com/maxeaubrey/aff612ba133fbfbf28f86c4820edbfd5

Atemu commented:
Found the error, fix incoming.

vcunat commented:
For reference: https://github.com/NixOS/nixpkgs/commit/1d19eeefeea4cb51d96592712acc4443c42efd0b
